infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/GatewayRouterProvisioner.java

```diff
@@ -0,0 +1,75 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.Executor;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.model.impl.ServerConfigImpl;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.environment.GatewayRouteConfig;
+import org.eclipse.che.api.workspace.server.spi.environment.InternalEnvironment;
+import org.eclipse.che.api.workspace.server.spi.provision.InternalEnvironmentProvisioner;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespace;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.external.GatewayRouteConfigGenerator;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.external.TraefikGatewayRouteConfigGenerator;
+import org.eclipse.che.workspace.infrastructure.kubernetes.util.KubernetesSharedPool;
+
+@Singleton
+public class GatewayRouterProvisioner {
+
+  protected static final Map<String, String> GATEWAY_CONFIGMAP_LABELS =
+      ImmutableMap.<String, String>builder()
+          .put("app", "che")
+          .put("role", "gateway-config")
+          .build();
+
+  private final KubernetesClientFactory clientFactory;
+  private final Executor k8sExecutor;
+
+  @Inject
+  public GatewayRouterProvisioner(KubernetesClientFactory clientFactory,
+      KubernetesSharedPool sharedPool) {
+    this.clientFactory = clientFactory;
+    this.k8sExecutor = sharedPool.getExecutor();
+  }
+
+  public List<ConfigMap> provision(RuntimeIdentity id, InternalEnvironment internalEnvironment)
+      throws InfrastructureException {
+    if (internalEnvironment.getGatewayRouteConfigs().isEmpty()) {
+      return Collections.emptyList();
+    }
+    List<ConfigMap> routeConfigMaps = new ArrayList<>();
+    KubernetesNamespace cheNamespace = new KubernetesNamespace(clientFactory, k8sExecutor, "che",
+        id.getWorkspaceId());
+
+    for (GatewayRouteConfig routeConfig : internalEnvironment.getGatewayRouteConfigs()) {
+      GatewayRouteConfigGenerator gatewayRouteConfigGenerator = new TraefikGatewayRouteConfigGenerator(
+          id.getInfrastructureNamespace());
+      gatewayRouteConfigGenerator.addRouteConfig(routeConfig);
+      ConfigMapBuilder configMapBuilder = new ConfigMapBuilder()
+          .withNewMetadata()
+          .withName(id.getWorkspaceId() + routeConfig.getName())
+          .withLabels(GATEWAY_CONFIGMAP_LABELS)
+          .withAnnotations(routeConfig.getAnnotations())
+          .endMetadata()
+          .withData(gatewayRouteConfigGenerator.generate());
+
+      ConfigMap routeConfigMap = cheNamespace.configMaps().create(configMapBuilder.build());
```

Comment:
I afraid you will face the permission problem accessing che namespace, which i have faced during search of CA configmap.

Reply:
good point. I'm afraid we would have to somehow create that configmap in namespace with gateway pod. If we won't have permissions for that, we would have to add it. I'll look what permissions we have and if we need more. Alternatively, we could configure configbump to watch configmaps on all namespaces, or namespaces with running workspaces.

Reply:
Would it be a big problem to require the che SA to be able to crud configmaps in the che namespace?

Reply:
it works with current helm install (we use `cluster-admin`). With operator, we currently can't use different namespaces anyway because lack of permissions and in one namespace it works fine.

So last option is OpenShift OAuth. I didn't try it, but this task is about Kubernetes so I would leave this potential permissions issue to https://github.com/eclipse/che/issues/17061